ksqldb-engine/src/main/java/io/confluent/ksql/internal/StorageUtilizationMetrics.java

```diff
@@ -0,0 +1,254 @@
+    if (registeredNodeMetrics.getAndSet(true)) {
+      return;
+    }
+    final MetricName nodeAvailable =
+        metricRegistry.metricName("node-storage-available", METRIC_GROUP);
+    final MetricName nodeTotal =
+        metricRegistry.metricName("node-storage-total", METRIC_GROUP);
+    final MetricName nodeUsed =
+        metricRegistry.metricName("node-storage-used", METRIC_GROUP);
+
+    metricRegistry.addMetric(
+        nodeAvailable,
+        (Gauge<Long>) (config, now) -> baseDir.getFreeSpace()
+    );
+    metricRegistry.addMetric(
+        nodeTotal,
+        (Gauge<Long>) (config, now) -> baseDir.getTotalSpace()
+    );
+    metricRegistry.addMetric(
+        nodeUsed,
+        (Gauge<Long>) (config, now) -> (baseDir.getTotalSpace() - baseDir.getFreeSpace())
+    );
+  }
+
+  @Override
+  public void metricChange(final KafkaMetric metric) {
+    if (!metric.metricName().name().equals("total-sst-files-size")) {
+      return;
+    }
+    final String taskId = metric.metricName().tags().getOrDefault("task-id", "");
+    final String queryId = getQueryId(metric);
+
+    // if we haven't seen a task for this query yet
+    synchronized (metricsSeen) {
+      if (!metricsSeen.containsKey(queryId)) {
+        metricRegistry.addMetric(
+            metricRegistry.metricName(
+            "query-storage-usage",
+              METRIC_GROUP,
+            ImmutableMap.of("query-id", queryId)),
+            (Gauge<BigInteger>) (config, now) -> computeQueryMetric(queryId)
+        );
+        metricsSeen.put(queryId, new HashMap<>());
+      }
+      final TaskStorageMetric newMetric;
+      // We haven't seen a metric for this query's task before
+      if (!metricsSeen.get(queryId).containsKey(taskId)) {
+        // create a new task level metric to track state store storage usage
+        newMetric = new TaskStorageMetric(
+          metricRegistry.metricName(
+            "task-storage-usage",
+            METRIC_GROUP,
+            ImmutableMap.of("task-id", taskId, "query-id", queryId)
+          ));
+        // add to list of seen task metrics for this query
+        metricsSeen.get(queryId).put(taskId, newMetric);
+        // create gauge for task level storage usage
+        metricRegistry.addMetric(
+            newMetric.metricName,
+            (Gauge<BigInteger>) (config, now) -> newMetric.getValue()
+        );
+      } else {
+        // We have this metric already
+        newMetric = metricsSeen.get(queryId).get(taskId);
+      }
+      newMetric.add(metric);
+    }
+  }
+
+  @Override
+  public void metricRemoval(final KafkaMetric metric) {
+    final MetricName metricName = metric.metricName();
+    if (!metricName.name().equals("total-sst-files-size")) {
+      return;
+    }
+
+    final String queryId = getQueryId(metric);
+    final String taskId = metric.metricName().tags().getOrDefault("task-id", "");
+    final TaskStorageMetric taskMetric = metricsSeen.get(queryId).get(taskId);
```

Comment:
move everything below this into another method like:

`private synchronized void handleRemovedSstFileSizeMetric(...`